# Mini Project2
### 20181531 김연주

In [1]:
import numpy as np
import pandas as pd

df = pd.read_csv("csv/train.csv")
df

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
0              0.288585          -0.020294          -0.132905   
1              0.278419          -0.016411          -0.123520   
2              0.279653          -0.019467          -0.113462   
3              0.279174          -0.026201          -0.123283   
4              0.276629          -0.016570          -0.115362   
...                 ...                ...                ...   
7347           0.299665          -0.057193          -0.181233   
7348           0.273853          -0.007749          -0.147468   
7349           0.273387          -0.017011          -0.045022   
7350           0.289654          -0.018843          -0.158281   
7351           0.351503          -0.012423          -0.203867   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
0            -0.995279         -0.983111         -0.913526         -0.995112   
1            -0.998245         -0.975300         -0.960322         -0.998807   
2            -0.995380         -0.967187         -0.978944         -0.996520   
3            -0.996091         -0.983403         -0.990675         -0.997099   
4            -0.998139         -0.980817         -0.990482         -0.998321   
...                ...               ...               ...               ...   
7347         -0.195387          0.039905          0.077078         -0.282301   
7348         -0.235309          0.004816          0.059280         -0.322552   
7349         -0.218218         -0.103822          0.274533         -0.304515   
7350         -0.219139         -0.111412          0.268893         -0.310487   
7351         -0.269270         -0.087212          0.177404         -0.377404   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
0            -0.983185         -0.923527         -0.934724  ...   
1            -0.974914         -0.957686         -0.943068  ...   
2            -0.963668         -0.977469         -0.938692  ...   
3            -0.982750         -0.989302         -0.938692  ...   
4            -0.979672         -0.990441         -0.942469  ...   
...                ...               ...               ...  ...   
7347          0.043616          0.060410          0.210795  ...   
7348         -0.029456          0.080585          0.117440  ...   
7349         -0.098913          0.332584          0.043999  ...   
7350         -0.068200          0.319473          0.101702  ...   
7351         -0.038678          0.229430          0.269013  ...   

      fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
0                           -0.710304                    -0.112754   
1                           -0.861499                     0.053477   
2                           -0.760104                    -0.118559   
3                           -0.482845                    -0.036788   
4                           -0.699205                     0.123320   
...                               ...                          ...   
7347                        -0.880324                    -0.190437   
7348                        -0.680744                     0.064907   
7349                        -0.304029                     0.052806   
7350                        -0.344314                    -0.101360   
7351                        -0.740738                    -0.280088   

      angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
0                                 0.030400                         -0.464761   
1                                -0.007435                         -0.732626   
2                                 0.177899                          0.100699   
3                                -0.012892                          0.640011   
4                                 0.122542                          0.693578   
...                                    ...                               ...   
7347                              0.829718                

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7352 entries, 0 to 7351
Columns: 563 entries, tBodyAcc-mean()-X to Activity
dtypes: float64(561), int64(1), object(1)
memory usage: 31.6+ MB


In [3]:
from sklearn.utils import shuffle
df = shuffle(df)

In [4]:
df

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
5542           0.277840          -0.017918          -0.111828   
3271           0.286987          -0.009430          -0.098485   
360            0.276053          -0.015409          -0.131719   
4527           0.132633          -0.056435          -0.248956   
5524           0.270431          -0.017142          -0.102501   
...                 ...                ...                ...   
3143           0.278094          -0.019110          -0.111255   
4608           0.266606          -0.019494          -0.107841   
237           -0.066239           0.079229          -1.000000   
1302           0.286715           0.000756          -0.062712   
2307           0.274016          -0.051143          -0.142694   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
5542         -0.995645         -0.993756         -0.977473         -0.996194   
3271         -0.996612         -0.981827         -0.986981         -0.996831   
360          -0.957768         -0.878283         -0.874007         -0.966534   
4527          0.122696          0.345314         -0.253909          0.072395   
5524         -0.996208         -0.997446         -0.991789         -0.996763   
...                ...               ...               ...               ...   
3143         -0.955781         -0.953467         -0.976319         -0.953478   
4608         -0.992436         -0.993846         -0.996191         -0.994043   
237          -0.683779         -0.538513         -0.085060         -0.687890   
1302         -0.017612          0.200435         -0.207874         -0.065327   
2307         -0.449426         -0.094656          0.000189         -0.472406   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
5542         -0.993720         -0.975025         -0.942488  ...   
3271         -0.981238         -0.987968         -0.938719  ...   
360          -0.883526         -0.863435         -0.900447  ...   
4527          0.350815         -0.283163          0.337778  ...   
5524         -0.996878         -0.991797         -0.939951  ...   
...                ...               ...               ...  ...   
3143         -0.947600         -0.974223         -0.916897  ...   
4608         -0.992985         -0.997009         -0.932105  ...   
237          -0.607355         -0.027535         -0.671284  ...   
1302          0.154043         -0.201844          0.126121  ...   
2307         -0.099923         -0.020736         -0.239664  ...   

      fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
5542                        -0.825305                     0.088311   
3271                        -0.760900                     0.018079   
360                         -0.298553                     0.127958   
4527                        -0.594570                     0.268471   
5524                        -0.525245                    -0.033310   
...                               ...                          ...   
3143                        -0.828059                     0.185925   
4608                        -0.834955                     0.023997   
237                         -0.445598                     0.050623   
1302                        -0.380916                     0.075615   
2307                        -0.542834                    -0.215645   

      angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
5542                             -0.324053                         -0.123347   
3271                              0.411057                          0.888819   
360                               0.160220                         -0.939252   
4527                             -0.394323                         -0.909823   
5524                              0.062593                          0.125413   
...                                    ...                               ...   
3143                              0.062161                

In [4]:
from sklearn.preprocessing import LabelEncoder

df_cat = df['Activity']
encoder = LabelEncoder()
encoder.fit(df_cat)
df_encoded = encoder.transform(df_cat)

In [5]:
df['Activity'] = df_encoded
df['Activity']

3852    1
548     2
3879    0
6388    3
3655    1
       ..
3001    4
1950    1
2522    5
2808    3
2358    5
Name: Activity, Length: 7352, dtype: int32

In [6]:
X = df.drop(['subject', 'Activity'], axis=1)
y = df['Activity']

## ExtraTreesClassifier

In [8]:
from sklearn.ensemble import ExtraTreesClassifier

extra_clf = ExtraTreesClassifier(random_state=42)
extra_clf.fit(X, y)

ExtraTreesClassifier(random_state=42)

In [9]:
from sklearn.model_selection import cross_val_score

extra_score = cross_val_score(extra_clf, X, y, cv=10, scoring="accuracy") # 교차검증
extra_score

array([0.99320652, 0.98641304, 0.98231293, 0.98911565, 0.99047619,
       0.98639456, 0.98911565, 0.98095238, 0.99591837, 0.98503401])

In [10]:
extra_score.mean()

0.9878939293108548

## Models

In [11]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecistionTreeClassifier

log_clf = LogisticRegression(solver="liblinear", random_state=42,max_iter=500)
rnd_clf = RandomForestClassifier(n_estimators=100, random_state=42)
svm_clf = SVC(kernel="linear", probability=True, random_state=42)
ada_clf = AdaBoostClassifier(RandomForestClassifier(random_state=42))
de_ada_clf = AdaBoostClassifier(DecistiongTreeClassifier(random_state=42))
gbrt_clf = GradientBoostingClassifier(random_state=42)
bag_clf = BaggingClassifier(
    RandomForestClassifier(random_state=42), bootstrap=False, random_state=42)

In [5]:
def display_scores(scores):
    print("Scores:", scores)
    print("Mean:", scores.mean())
    print("Standard deviation", scores.std())

In [13]:
log_clf.fit(X, y)
rnd_clf.fit(X, y)
svm_clf.fit(X, y)
ada_clf.fit(X, y)
gbrt_clf.fit(X, y)
bag_clf.fit(X, y)

BaggingClassifier(base_estimator=RandomForestClassifier(random_state=42),
                  bootstrap=False, random_state=42)

In [14]:
log_score = cross_val_score(log_clf, X, y, cv=10, scoring="accuracy")
display_scores(log_score)

array([0.98913043, 0.98097826, 0.99183673, 0.98639456, 0.9877551 ,
       0.98095238, 0.9877551 , 0.98503401, 0.9877551 , 0.98367347])

In [16]:
rnd_score = cross_val_score(rnd_clf, X, y, cv=10, scoring="accuracy")
display_scores(rnd_score)

array([0.98641304, 0.98777174, 0.98095238, 0.97959184, 0.98095238,
       0.98639456, 0.98367347, 0.98367347, 0.98503401, 0.97823129])

In [18]:
svm_score = cross_val_score(svm_clf, X, y, cv=10, scoring="accuracy")
display_scores(svm_score)

array([0.99048913, 0.98913043, 0.99183673, 0.98639456, 0.9877551 ,
       0.9877551 , 0.9877551 , 0.9877551 , 0.9877551 , 0.99047619])

In [20]:
ada_clf_score = cross_val_score(ada_clf, X, y, cv=10, scoring="accuracy")
display_scores(ada_clf_score)

array([0.98777174, 0.98641304, 0.97959184, 0.97959184, 0.97959184,
       0.98639456, 0.98231293, 0.98231293, 0.9877551 , 0.97687075])

In [49]:
from sklearn.tree import DecisionTreeClassifier
de_ada_clf = AdaBoostClassifier(DecisionTreeClassifier(random_state=42), random_state=42)
de_ada_clf.fit(X,y)

AdaBoostClassifier(base_estimator=DecisionTreeClassifier(random_state=42))

In [50]:
de_ada_clf_score = cross_val_score(de_ada_clf, X, y, cv=10, scoring="accuracy")
display_scores(de_ada_clf_score)

array([0.95516304, 0.91983696, 0.95238095, 0.93605442, 0.93741497,
       0.95238095, 0.92517007, 0.94285714, 0.93605442, 0.95102041])

In [23]:
from sklearn.ensemble import ExtraTreesClassifier

extra_trees_clf = ExtraTreesClassifier(n_estimators=50, random_state=42)
extra_trees_clf.fit(X, y)

ExtraTreesClassifier(n_estimators=50, random_state=42)

In [24]:
from sklearn.model_selection import cross_val_score

extra_fold_score = cross_val_score(extra_trees_clf, X, y, cv=10, scoring="accuracy")
display_scores(extra_fold_score)

array([0.99048913, 0.98369565, 0.98639456, 0.9877551 , 0.98095238,
       0.9877551 , 0.98639456, 0.98367347, 0.99183673, 0.98231293])

## Voting

In [26]:
voting_clf = VotingClassifier(
    estimators=[('lr', log_clf), ('rf', rnd_clf), ('svc', svm_clf)],
    voting='hard')

In [27]:
voting_clf.fit(X, y)

VotingClassifier(estimators=[('lr',
                              LogisticRegression(random_state=42,
                                                 solver='liblinear')),
                             ('rf', RandomForestClassifier(random_state=42)),
                             ('svc',
                              SVC(kernel='linear', probability=True,
                                  random_state=42))])

In [28]:
from sklearn.model_selection import cross_val_score

voting_clf_score = cross_val_score(voting_clf, X, y, cv=10, scoring="accuracy")
display_scores(voting_clf_score)

array([0.98913043, 0.98777174, 0.99183673, 0.9877551 , 0.99047619,
       0.98639456, 0.99183673, 0.9877551 , 0.98911565, 0.9877551 ])

In [6]:
display_scores(voting_clf_score)

NameError: name 'voting_clf_score' is not defined

In [40]:
log_clf = LogisticRegression(solver="liblinear", random_state=42,max_iter=1000)
voting_clf1 = VotingClassifier(
    estimators=[('lr', log_clf),('rf', rnd_clf), ('svc', svm_clf)],voting='hard')
voting_clf1.fit(X, y)

VotingClassifier(estimators=[('lr',
                              LogisticRegression(max_iter=1000, random_state=42,
                                                 solver='liblinear')),
                             ('rf', RandomForestClassifier(random_state=42)),
                             ('svc', SVC(probability=True, random_state=42))])

In [41]:
from sklearn.model_selection import cross_val_score

voting1_clf_score = cross_val_score(voting_clf1, X, y, cv=10, scoring="accuracy")
display_scores(voting1_clf_score)

array([0.98777174, 0.98233696, 0.9877551 , 0.98503401, 0.98503401,
       0.98095238, 0.98367347, 0.98367347, 0.98639456, 0.98231293])

In [ ]:
display_scores(voting_clf_score)

- LogisticRegression: 0.9861265158237208
- RandomForestClassifier: 0.9823177314404022
- SVC: 0.988710255841467
- AdaBoostClassifier: 0.9805492088139603
- GradientBoostingClassifier:0.9865341245193731 
- BaggingClassifier: 0.9831340579710144

### SVM의 결과가 제일 좋게 나왔음

## 결과: 0.988710255841467 => 약 98.9%